zyte_common_items/pages.py

```diff
@@ -1,35 +1,96 @@
 from datetime import datetime
+from typing import Generic, Type, TypeVar
 
 import attrs
 from web_poet import ItemPage, RequestUrl, Returns, WebPage, field
 from web_poet.pages import ItemT
 
-from .components import Metadata
-from .items import Article, ArticleList, BusinessPlace, Product, ProductList, RealEstate
-from .util import format_datetime
+from .components import (
+    ArticleListMetadata,
+    ArticleMetadata,
+    BusinessPlaceMetadata,
+    ProductListMetadata,
+    ProductMetadata,
+    ProductNavigationMetadata,
+    RealEstateMetadata,
+)
+from .items import (
+    Article,
+    ArticleList,
+    BusinessPlace,
+    Product,
+    ProductList,
+    ProductNavigation,
+    RealEstate,
+)
+from .util import format_datetime, metadata_processor
+
+try:
+    from typing import get_args as _get_args
+except ImportError:
+
+    def _get_args(base):  # type: ignore
+        return getattr(base, "__args__", ())
+
+
+#: Generic type for metadata classes for specific item types.
+MetadataT = TypeVar("MetadataT")
 
 
 def _date_downloaded_now():
     return format_datetime(datetime.utcnow())
 
 
-class _BasePage(ItemPage[ItemT]):
+class HasMetadata(Generic[MetadataT]):
+    """Inherit from this generic mixin to set the metadata class used by a page
+    class."""
+
+    @property
+    def metadata_cls(self) -> Type[MetadataT]:
+        """Metadata class."""
+        return _get_metadata_class(self)
+
+
+def _get_metadata_class(obj):
+    for base in getattr(obj.__class__, "__orig_bases__", []):
+        origin = getattr(base, "__origin__", None)
+        if not origin:
+            continue
+        if origin != HasMetadata:
+            continue
+        return _get_args(base)[0]
+    return None
```

Comment:
With an implementation like this is makes sense to check that the inheritance works properly, e.g. a page object subclass which wants to use different metadata.

Reply:
I've added a test, can you please check that it tests what you wanted?

Reply:
In the test the subclass uses the same metadata class (ProductMetadata). I think it makes sense to test that it's possible to change the metadata class used in a subclass, e.g. use CustomProductMetadata in a ProductPage subclass.

Reply:
Nevermind; I see now that the very next test checks exactly this :)